### DISCUSSION TITLE
bot starts but doesn't login

### DISCUSSION THREAD

**Discussion by dolphin-cat** [PrismarineJS/mineflayer#2182] [Q&A] [answered]
code:
```py
from javascript import require, On
mineflayer = require('mineflayer')
import random, time
RANGE_GOAL = 1
BOT_USERNAME = '***********'
bot = mineflayer.createBot({
  'host': '************.aternos.me',
  'port': 25565,
  'username': BOT_USERNAME
})
print("Started mineflayer")
@On(bot, 'spawn')
def handle(*args):
  print("I spawned 👋")
  mcData = require('minecraft-data')(bot.version)
  movements = pathfinder.Movements(bot, mcData)
@On(bot, "chat")
def handle(*args):
  bot.chat('/login *************')
  bot.setControlState('forward', true)
  time.sleep(3)
  bot.setControlState('forward', false)
  while True:
    r1 = random.randint(1, 9)
    r2 = random.randint(1, 9)
    r3 = random.randint(1, 9)
    r4 = random.randint(1, 9)
    r5 = random.randint(1, 9)
    rnums = r1 + r2 + r3 + r4 + r5
    bot.chat('test' + rnums)
    time.sleep(3)
@On(bot, "end")
def handle(*args):
  print("Bot ended!")
  exit()
```
output:
D:\bot>python bot.py
Started mineflayer
Bot ended!

**Comment by WhoTho** [comment]
Hey,

The command for on("event", function) command is not "@on(bot, function)", it is "bot.(event, function)". For example,

bot.on("spawn", () => { bot.chat("I spawned")})

This will be activated every time the bot spawns, either when logging in or respawning. If you want something to happen once, use bot.once.

**Comment by extremeheat** [comment]
> The command for on("event", function) command is not "@on(bot, function)", it is "bot.(event, function)". For example,

this code is in python

> Bot ended!

Your bot is getting disconnected. You need to log the kicked and error events:
```py
bot.on('kicked', print)
bot.on('error', print)
```

to figure out why

**Comment by dolphin-cat** [comment]
[JSE] (node:10216) UnhandledPromiseRejectionWarning: Error: Attempt to access '' failed. Python didn't respond in time (100000ms), look above for any Python errors. If no errors, the API call hung.
[JSE]     at D:\Anaconda3\lib\site-packages\javascript\js\pyi.js:149:13
[JSE]     at waitFor (D:\Anaconda3\lib\site-packages\javascript\js\pyi.js:46:26)
[JSE]     at async PyBridge.call (D:\Anaconda3\lib\site-packages\javascript\js\pyi.js:148:18)
[JSE] (Use `node --trace-warnings ...` to show where the warning was created)
[JSE] (node:10216) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
[JSE] (node:10216) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Comment by extremeheat** [comment]
Cannot reproduce. 

Try connecting to the prismarinejs test server:

bot = mineflayer.createBot({ 'host': '95.111.249.143', 'port': 10000, 'username': BOT_USERNAME, 'hideErrors': False })

**Comment by dolphin-cat** [comment]
is it running in offline mode? cuz my name is cracked, not a paid mc acc

**Comment by dolphin-cat** [comment]
it is still open

**Comment by dolphin-cat** [comment]
Started mineflayer
Error in exception handler
[Errno 2] No such file or directory: 'events.js'
** JavaScript Stacktrace **
TypeError [ERR_INVALID_ARG_TYPE]: The "listener" argument must be of type function. Received undefined
    at checkListener (events.js:112:11)
    at _addListener (events.js:348:3)
    at EventEmitter.addListener (events.js:406:10)
    at Bridge.call (D:\Anaconda3\lib\site-packages\javascript\js\bridge.js:135:42)
    at Bridge.onMessage (D:\Anaconda3\lib\site-packages\javascript\js\bridge.js:218:25)
    at Socket.<anonymous> (D:\Anaconda3\lib\site-packages\javascript\js\bridge.js:267:14)
    at Socket.emit (events.js:315:20)
    at addChunk (internal/streams/readable.js:309:12)
    at readableAddChunk (internal/streams/readable.js:284:9)
    at Socket.Readable.push (internal/streams/readable.js:223:10)
** Python Stacktrace **
  File "bot.py", line 10, in <module>
    bot.on('spawn')

  File "D:\Anaconda3\lib\site-packages\javascript\proxy.py", line 208, in __call__
    else self._exe.callProp(

  File "D:\Anaconda3\lib\site-packages\javascript\proxy.py", line 152, in callProp
    resp = self.pcall(ffid, "call", method, args, timeout=timeout, forceRefs=forceRefs)

  File "D:\Anaconda3\lib\site-packages\javascript\proxy.py", line 140, in pcall
    raise JavaScriptError(attr, res["error"])

**Comment by extremeheat** [comment, marked as answer]
Please see the tutorial linked in the README or click on the Open in Colab badge. This will teach you some of the basics.
